WS17_18/04_tableView/your assignment/muesse/Yatodoa/Controller/TodoTableViewController.swift

```diff
@@ -0,0 +1,172 @@
+        return lhs.hashValue < rhs.hashValue
+    }
+}
+
+class TodoTableViewController: UITableViewController {
+
+    var todos: [ToDo] = [] {
+        didSet {
+            print(todos)
+            
+            todoSections = todos.groupBy(key: { todo in
+                return TodoType(todo: todo)
+            }).sorted(by: { (l, r) in
+                return l.key < r.key
+            })
+        }
+    }
+    
+    var todoSections: [(key: TodoType, value: [ToDo])] = []
+    
+    override func viewDidLoad() {
+        super.viewDidLoad()
+        
+        tableView.register(UINib(nibName: TodoTableViewCell.NibName, bundle: nil), forCellReuseIdentifier: TodoTableViewCell.ReuseIdentifier)
+        tableView.register(UINib(nibName: AddTodoTableViewCell.NibName, bundle: nil), forCellReuseIdentifier: AddTodoTableViewCell.ReuseIdentifier)
+       
+        todos.append(ToDo.empty)
+    }
+
+    // MARK: - Table view data source
+
+    override func numberOfSections(in tableView: UITableView) -> Int {
+        return todoSections.count
+    }
+
+    override func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
+        return todoSections[section].value.count
+    }
+
+    override func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
+        switch todoSections[indexPath.section].key {
+        case .add:
+            let cell = tableView.dequeueReusableCell(withIdentifier: AddTodoTableViewCell.ReuseIdentifier, for: indexPath) as! AddTodoTableViewCell
+            
+            cell.delegate = self
+            
+            return cell
+        case .open, .done:
+            let cell = tableView.dequeueReusableCell(withIdentifier: TodoTableViewCell.ReuseIdentifier, for: indexPath) as! TodoTableViewCell
+            
+            let todo = todoSections[indexPath.section].value[indexPath.row]
+            cell.todo = todo.task
+            cell.completed = todo.completed
+            cell.delegate = self
+            
+            return cell
+        }
+    }
+    
+    override func tableView(_ tableView: UITableView, titleForHeaderInSection section: Int) -> String? {
+        return todoSections[section].key.title
+    }
+
+    override func tableView(_ tableView: UITableView, shouldHighlightRowAt indexPath: IndexPath) -> Bool {
+        return false
+    }
+    
+    override func tableView(_ tableView: UITableView, canEditRowAt indexPath: IndexPath) -> Bool {
+        switch todoSections[indexPath.section].key {
+        case .add, .done:
+            return false
+        case .open:
+            return true
+        }
+    }
+    
+    override func tableView(_ tableView: UITableView, editingStyleForRowAt indexPath: IndexPath) -> UITableViewCellEditingStyle {
+        return .delete
+    }
+    
+    override func tableView(_ tableView: UITableView, commit editingStyle: UITableViewCellEditingStyle, forRowAt indexPath: IndexPath) {
```

Comment:
Der Zugriff über `indexPath` auf das "normale Model" bringt einige Gefahren mit sich. So funktioniert die Implementierung beispielsweise nur, wenn es bei zwei Sections verbleibt. Kommt die "done" Section hinzu, so crashed die App. 

Man sollte (fast) immer mit dem indexPath auf das dedizierte Array, sprich `todoSections` zugreifen, weil diese ein **1 zu 1** mapping haben, wie in `private func insert(_ todo: ToDo, animated: Bool)`.

In der Implementierung von #18 ist solch ein Zugriff zu sehen.